core/capabilities/webapi/trigger.go

```diff
@@ -1,18 +1,268 @@
+	AllowedTopics  []string                 `toml:"allowedTopics"`
+	RateLimiter    common.RateLimiterConfig `toml:"rateLimiter"`
+	// RequiredParams is advisory to the consumer, it is not enforced.
+	RequiredParams []string `toml:"requiredParams"`
+}
+
+type webapiTrigger struct {
+	allowedSenders map[string]bool
+	allowedTopics  map[string]bool
+	ch             chan<- capabilities.TriggerResponse
+	config         TriggerConfig
+	rateLimiter    *common.RateLimiter
+}
+
+type triggerConnectorHandler struct {
+	services.StateMachine
+
+	capabilities.CapabilityInfo
+	capabilities.Validator[TriggerConfig, Input, capabilities.TriggerResponse]
+	connector           connector.GatewayConnector
+	lggr                logger.Logger
+	mu                  sync.Mutex
+	registeredWorkflows map[string]webapiTrigger
+}
+
+var _ capabilities.TriggerCapability = (*triggerConnectorHandler)(nil)
+var _ services.Service = &triggerConnectorHandler{}
+
+func NewTrigger(config string, registry core.CapabilitiesRegistry, connector connector.GatewayConnector, lggr logger.Logger) (*triggerConnectorHandler, error) {
+	if connector == nil {
+		return nil, errors.New("missing connector")
+	}
+	handler := &triggerConnectorHandler{
+		Validator:           capabilities.NewValidator[TriggerConfig, Input, capabilities.TriggerResponse](capabilities.ValidatorArgs{Info: webapiTriggerInfo}),
+		connector:           connector,
+		registeredWorkflows: map[string]webapiTrigger{},
+		lggr:                lggr.Named("WorkflowConnectorHandler"),
+	}
+
+	return handler, nil
+}
+
+// processTrigger iterates over each topic, checking against senders and rateLimits, then starting event processing and responding
+func (h *triggerConnectorHandler) processTrigger(ctx context.Context, gatewayID string, body *api.MessageBody, sender ethCommon.Address, payload webapicapabilities.TriggerRequestPayload) {
+	// Pass on the payload with the expectation that it's in an acceptable format for the executor
+	wrappedPayload, err := values.WrapMap(payload)
+	if err != nil {
+		h.lggr.Errorw("Error wrapping payload")
+		h.sendResponse(ctx, gatewayID, body, webapicapabilities.TriggerResponsePayload{Status: "ERROR", ErrorMessage: "Error wrapping payload"})
+		return
+	}
+	matchedWorkflows := 0
+	var response webapicapabilities.TriggerResponsePayload
+
+	for _, trigger := range h.registeredWorkflows {
+		topics := payload.Topics
+		// empty topics means all topics
+		if len(topics) == 0 {
+			for k := range trigger.allowedTopics {
+				topics = append(topics, k)
+			}
+		}
+
+		for _, topic := range topics {
+			if trigger.allowedTopics[topic] {
+				matchedWorkflows++
+
+				if !trigger.allowedSenders[sender.String()] {
+					h.lggr.Errorw("Unauthorized Sender", "sender", sender.String(), "messageID", body.MessageId)
+					h.sendResponse(ctx, gatewayID, body, webapicapabilities.TriggerResponsePayload{Status: "ERROR", ErrorMessage: "Unauthorized Sender"})
+					return
```

Comment:
As an illustration of why iterating over every workflow seems off:

I send a request with no topics, as sender 0x1:
```
  {
    jsonrpc: "2.0",
    id: "...",
    method: "web-trigger",
    params: {
      signature: "...",
      body: {
        don_id: "workflow_123",
        payload: {
          trigger_id: "web-trigger@1.0.0",
          trigger_event_id: "action_1234567890",
          timestamp: 1234567890,
          topics: [],
          params: {}
        }
      }
    }
  }
```
On the first iteration because there are no topics, it fills itself on line 93 with the topics of the first trigger.
It passes through the if check on line 99 because it inherited all of the topics.
But now on line 102 if the first trigger does have address 0x1 as an allowed sender it will fail.

So no one will be able to use empty topics. And there are conflicts in multiple users using the same topic.

Put another way: topics are scoped too high and conflicting

Reply:
I think empty topics should be removed for this and other complexities.  I think a consumer should explicitly know which triggers to invoke.